tls/extensions/s2n_extension_type.c

```diff
@@ -151,6 +151,26 @@ int s2n_extension_recv(const s2n_extension_type *extension_type, struct s2n_conn
     return S2N_SUCCESS;
 }
 
+int s2n_extension_is_missing(const s2n_extension_type *extension_type, struct s2n_connection *conn)
+{
+    notnull_check(extension_type);
+    notnull_check(extension_type->if_missing);
+    notnull_check(conn);
+
+    s2n_extension_type_id extension_id;
+    GUARD(s2n_extension_supported_iana_value_to_id(extension_type->iana_value, &extension_id));
+
+    /* Do not consider an extension missing if we did not send a request */
+    if(extension_type->is_response &&
+            !S2N_CBIT_TEST(conn->extension_requests_sent, extension_id)) {
+        return S2N_SUCCESS;
+    }
```

Comment:
Q: Depending on server/client mode, extensions may be received prior to being sent or vice versa. 
Do we want to call an extension not-missing if we have sent, not yet received and is_response is set? How do we determine the order in which the extension is sent/recv?

Reply:
This is the relevant part of the RFC:
```
   Implementations MUST NOT send extension responses if the remote
   endpoint did not send the corresponding extension requests, with the
   exception of the "cookie" extension in the HelloRetryRequest.  Upon
   receiving such an extension, an endpoint MUST abort the handshake
   with an "unsupported_extension" alert.
```
Basically, most extensions have a request and a response form. For example, the client requests a max_frag_length in one extensions and the server sends a response to confirm that max_frag_length in a different extension. We shouldn't send or receive the response if we never sent or received the request.

But by the same token, we shouldn't call "if_missing" for a response extension we never sent the request for. If the response is actually mandatory, we should always send the request :)